DE_pipeline.snake

```diff
@@ -687,96 +814,96 @@ rule annotation:
-			contrast_def = lfc_def = 'contrast=c("{}", "{}", "{}")'.format(contrast_column, contrast_num, contrast_ref)
-		elif "vector" in contrast:
-			contrast_type          = "vector"
-			contrast_def = lfc_def = 'contrast=c({})'.format(", ".join(map(str, contrast["vector"])))
-		else:
-			raise ValueError("Error in contrast: no valid contrast definition! must be one of {}".format(["coef:", "ratio:", "vector:"]))
-		
-		# parameters
-		cutoff_FDR           =           contrast["max_p_adj"]
-		lfcThreshold         =           contrast["results_parameters"]["lfcThreshold"]
-		altHypothesis        =           contrast["results_parameters"]["altHypothesis"]
-		independentFiltering = "TRUE" if contrast["results_parameters"]["independentFiltering"] else "FALSE"
-		lfc_shrink_type      =           contrast["lfcShrink_parameters"]["type"]
-		rank_by              =           contrast["ranking_by"]
-		rank_order           =           contrast["ranking_order"].replace("x", "res$"+rank_by)
-		
-		script = textwrap.dedent(r"""
-		#----- import packages
-		library(AnnotationDbi)
-		library(DESeq2)
-		library({params.annot_pkg})
-		
-		{R_SESSION_INFO}
-
-    
-		
-		#----- load DESeqDataSet
-		dds <- readRDS("{input}")
-		
-		#----- contrast
-		if ("{contrast_type}" == "ratio") {{
-			dds${contrast_column} = relevel(dds${contrast_column}, ref="{contrast_ref}")
-			dds <- nbinomWaldTest(dds)
-		}}
-		res <- results(dds, {contrast_def}, # alpha={cutoff_FDR}, lfcThreshold={lfcThreshold}, 
-		               altHypothesis="{altHypothesis}") #, independentFiltering={independentFiltering})
-
-		#----- log-fold-change shrinkage
-		#res$lfcShrunk   <- res$log2FoldChange
-		#res$lfcShrunkSE <- res$lfcSE
-		if ("{lfc_shrink_type}"!="none"){{
-			lfc <- lfcShrink(dds, {lfc_def}, # lfcThreshold={lfcThreshold}, 
-				type="{lfc_shrink_type}")
-			res$log2FoldChange_orig <- res$log2FoldChange
-			res$log2FoldChange      <- lfc$log2FoldChange
-			res$lfcSE_orig          <- res$lfcSE
-			res$lfcSE               <- lfc$lfcSE
-		}}
-		
-		#----- ranking
-		res <- res[order({rank_order}),]
-		
-		#----- annotation [BEWARE- the in case on multiple mapping, the first entry will be selected]
-		res$symbol <- mapIds({params.annot_pkg}, keys=sub("\\.[0-9]+$", "", row.names(res)), 
-		                     column="SYMBOL", keytype="ENSEMBL", multiVals="first")
-		res$entrez <- mapIds({params.annot_pkg}, keys=sub("\\.[0-9]+$", "", row.names(res)), 
-		                     column="ENTREZID", keytype="ENSEMBL", multiVals="first")
-		
-		#----- save as rds & csv
-		saveRDS(res, file="{output.rds}")
-		write.table(data.frame(gene_id=rownames(res), res), file="{output.csv}", sep="\t", col.names=TRUE, row.names=FALSE, quote=FALSE)
-		""")
-		
-		script_file = pph.log(log.out, snakemake_format(script), step="contrasts_full", extension="R", **wildcards)
-		shell("Rscript --vanilla '{script_file}' &>> '{log.out}'")
-
+# rule contrasts_full:
```

Comment:
I think that this is an issue of the type "what should we do about this part". I have initially included this rule to generate full contrasts with no pre-filtering because I think that tmod works better like this. However, I am still unsure whether to keep it. I think Eric commented this because this rule doubles a lot of computations.

Reply:
I looked again at the differences betweeen `contrasts` & `contrasts_full` rules. I can't see any meaningful differences except for the fact that `contrasts_full` doesn't use the user's choice set in the configuration for the significance level, log fold change cutoff and independent filtering.

```
< 		res <- results(dds, {contrast_def}, alpha={cutoff_FDR}, lfcThreshold={lfcThreshold}, 
< 		               altHypothesis="{altHypothesis}", independentFiltering={independentFiltering})
---
> 		res <- results(dds, {contrast_def}, # alpha={cutoff_FDR}, lfcThreshold={lfcThreshold}, 
> 		               altHypothesis="{altHypothesis}") #, independentFiltering={independentFiltering})
65c65,66
< 			lfc <- lfcShrink(dds, {lfc_def}, lfcThreshold={lfcThreshold}, type="{lfc_shrink_type}")
---
> 			lfc <- lfcShrink(dds, {lfc_def}, # lfcThreshold={lfcThreshold}, 
> 				type="{lfc_shrink_type}")
```

So when the `contrasts_full` rule is used, the `DESeq2` default options are used, instead of the corresponding values set in the configuration. I think it is potentially misleading. I think we rather should set the default values in the configuration to match those in DESeq2, and remove completely the `contrasts_full` rule. So when the user doesn't set these values in the configuration, the outcome is identical to what would have been computed by the `contrasts_full` rule.

This are the actions that I will take in the newest commit.

Reply:
As explained previously, if you remove it, tmod will not run correctly. Tmod relies on p-values computed for all genes, so if for a fraction of genes the p-values are not computed, their order remains as it was in the count file. If this order is not random, artifact may appear. Which was the behaviour I have observed which prompted me to create the contrasts_full rule.

I agree, however, that in contrasts_full the option independentFiltering should be set to FALSE to make absolutely clear what this is about.

Another solution would be to resample genes with no p-values assigned prior to running tmod, but I have not tested yet this approach.

Reply:
how about simply shuffling the gene order before running tmod? this would be more in line with what users expect, would save time and avoid redundant computation